ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java

```diff
@@ -202,6 +217,23 @@ private static LogicalSchema buildSchema(
     final String errorMsg = "Error computing new key from expression "
         + expressionMetadata.getExpression();
 
-    return row -> expressionMetadata.evaluate(row, null, logger, () -> errorMsg);
+    return new PartitionByExpressionEvaluator(
+        row -> expressionMetadata.evaluate(row, null, logger, () -> errorMsg),
```

Comment:
I suspect this isn't the only place where this bug happens (a null value doesn't get properly passed into an `ExpressionEvaluator` 🤔). I had a PR at one point going to try to have `ExpressionMetadata#evaluate` take in both a key and a value instead of just a `GenericRow` which would have prevented this issue. Unfortunately, I couldn't get joins to work because the joiner doesn't actually have access to the key 😭 

tldr, I think it would be good if we encapsulate as much of this "put the keys and only the keys into the generic row if the row was null" into the `PartitionByExpressionEvaluator` and have that implement `BiFunction<Struct, GenericRow, Object>` so at least the part that creates the key is all in the same place that calls the evaluator:
```java
      final Object newKey = evaluator.evaluator.apply(
          evaluator.acceptsKey
          ? GenericRow.fromList(StructKeyUtil.asList(k))
          : v
      );
```

I would also just pass in `expressionMetadata` into the `PartitionByExpressionEvaluator` instead of passing it in as a lambda so that it's centralized and we don't have to track down where it's being called and what's getting passed into it

Reply:
Ah, thanks for the suggestion -- much cleaner!

What's the benefit of implementing `BiFunction<Object, GenericRow, Object>` rather than simply have a method? I've currently got
```
  Object evaluate(final Object key, final GenericRow value) {
      final GenericRow row = evaluateOnKeyOnly
          ? GenericRow.fromList(StructKeyUtil.asList(key))
          : value;
      return expressionMetadata.evaluate(row, null, logger, errorMsg);
    }
```

Reply:
there is none, I think that's great (especially if nowhere else expects the BiFunction)